fs-storage/src/folder_storage.rs

```diff
@@ -0,0 +1,784 @@
+        Ok(())
+    }
+    /// Resolve differences between memory and disk data
+    fn resolve_divergence(&mut self) -> Result<()> {
+        let new_data = FolderStorage::new("new_data".into(), &self.path)?;
+
+        for (key, new_value) in new_data.data.iter() {
+            if let Some(existing_value) = self.data.get(key) {
+                let existing_value_updated = self
+                    .modified
+                    .get(key)
+                    .and_then(|ram_stamp| {
+                        self.written_to_disk
+                            .get(key)
+                            .map(|disk_stamp| ram_stamp > disk_stamp)
+                    })
+                    .unwrap_or(false);
+
+                // Use monoid to combine value for the given key
+                // if the memory and disk have diverged
+                if existing_value_updated {
+                    let resolved_value = V::combine(existing_value, new_value);
+                    self.data.insert(key.clone(), resolved_value);
+                } else {
+                    self.data.insert(key.clone(), new_value.clone());
+                }
+            } else {
+                self.data.insert(key.clone(), new_value.clone());
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Remove files from disk that are not present in memory
+    fn remove_files_not_in_ram(&mut self) -> Result<()> {
+        for key in self.deleted_keys.iter() {
+            let file_path = self.path.join(format!("{}.json", key));
+            if file_path.exists() {
+                fs::remove_file(&file_path).expect("Failed to delete file");
+            }
+        }
+        Ok(())
+    }
+
+    pub fn extract_key_from_file_path(&self, path: &Path) -> Result<K> {
+        path.file_stem()
+            .ok_or_else(|| {
+                ArklibError::Storage(
+                    self.label.clone(),
+                    "Failed to extract file stem from filename".to_owned(),
+                )
+            })?
+            .to_str()
+            .ok_or_else(|| {
+                ArklibError::Storage(
+                    self.label.clone(),
+                    "Failed to convert file stem to string".to_owned(),
+                )
+            })?
+            .parse::<K>()
+            .map_err(|_| {
+                ArklibError::Storage(
+                    self.label.clone(),
+                    "Failed to parse key from filename".to_owned(),
+                )
+            })
+    }
+}
+
+impl<K, V> BaseStorage<K, V> for FolderStorage<K, V>
+where
+    K: Ord
+        + Clone
+        + serde::Serialize
+        + serde::de::DeserializeOwned
+        + std::str::FromStr
+        + std::fmt::Display,
+    V: Clone + serde::Serialize + serde::de::DeserializeOwned + Monoid<V>,
+{
+    /// Set a key-value pair in the internal mapping
+    fn set(&mut self, key: K, value: V) {
+        self.data.insert(key.clone(), value);
+        self.deleted_keys.remove(&key);
```

Comment:
Just curious, why do we need to remove it from the deleted keys?

Reply:
`deleted_keys` is a `BTreeSet`, which just stores the key to be deleted. that's why we also clear it after sync. if it was BtreeMap, then we just compare it, just as we do with other `timestamps`.

Reply:
This is to ensure that deleted keys and keys in the mapping are a disjoint status. This helps us avoid iterating over keys and then filtering for non-deleted keys.

Reply:
Am I right, that we keep keys in `deleted_keys` until they are propagated to other devices ("soft delete")? And if we encounter same key introduced again, then we remove it from "soft delete"?

(Just guessing, haven't read the code yet)

Reply:
Yup that is how it is used